File: app/src/main/java/com/example/shercofaqapp/model/repositories/BikeRepository.java
package com.example.shercofaqapp.model.repositories;

import android.app.Application;
import android.os.AsyncTask;

import androidx.lifecycle.LiveData;

import com.example.shercofaqapp.model.Bike;
import com.example.shercofaqapp.model.BikeDao;
import com.example.shercofaqapp.model.MyBikesDatabase;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.ktx.Firebase;

import java.util.List;

public class BikeRepository {

    private BikeDao bikeDao;
    private LiveData<List<Bike>> bikes;

    public BikeRepository(Application application) {

        MyBikesDatabase myBikesDatabase = MyBikesDatabase.getInstance(application);
        bikeDao = myBikesDatabase.getBikeDao();

    }

    public LiveData<List<Bike>> getBikes() {

        return bikeDao.getAllBikes();

    }

    public Bike getBike(long id) {

        return bikeDao.getBike(id);

    }

    //Insert methods
    public void insertBike(Bike bike) {

        new InsertBikeAsyncTask(bikeDao).execute(bike);

    }

    private static class InsertBikeAsyncTask extends AsyncTask<Bike, Void, Void> {

        private BikeDao bikeDao;

        public InsertBikeAsyncTask(BikeDao bikeDao) {
            this.bikeDao = bikeDao;
        }

        @Override
        protected Void doInBackground(Bike... bikes) {

            bikeDao.insert(bikes[0]);


            return null;
        }
    }

    //Update methods
    public void updateBike(Bike bike) {

        new UpdateBikeAsyncTask(bikeDao).execute(bike);

    }

    private static class UpdateBikeAsyncTask extends AsyncTask<Bike, Void, Void> {

        private BikeDao bikeDao;

        public UpdateBikeAsyncTask(BikeDao bikeDao) {
            this.bikeDao = bikeDao;
        }

        @Override
        protected Void doInBackground(Bike... bikes) {

            bikeDao.update(bikes[0]);

            return null;
        }
    }

    //Delete methods
    public void deleteBike(Bike bike) {

        new DeleteBikeAsyncTask(bikeDao).execute(bike);

    }

    private static class DeleteBikeAsyncTask extends AsyncTask<Bike, Void, Void> {

        private BikeDao bikeDao;

        public DeleteBikeAsyncTask(BikeDao bikeDao) {
            this.bikeDao = bikeDao;
        }

        @Override
        protected Void doInBackground(Bike... bikes) {

            bikeDao.delete(bikes[0]);

            return null;
        }
    }

}

File: app/src/main/java/com/example/shercofaqapp/model/repositories/BikeRepository.kt
package com.example.shercofaqapp.model.repositories

import androidx.lifecycle.MutableLiveData
import com.example.shercofaqapp.model.Bike
import com.example.shercofaqapp.utils.*
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener

class BikeRepository {

    fun createBike(bike: Bike, bikeFirebaseKey: String) {

        REF_DATABASE_ROOT
            .child(USERS_NODE)
            .child(CURRENT_UID)
            .child(BIKES_NODE)
            .child(bikeFirebaseKey).setValue(bike)
    }

    fun readBikes(liveData: MutableLiveData<List<Bike>>) {
        val userId = FirebaseAuth.getInstance().currentUser?.uid

        REF_DATABASE_ROOT
            .child(USERS_NODE)
            .child(userId!!)
            .child(BIKES_NODE)
            .addValueEventListener(object : ValueEventListener {

            override fun onDataChange(snapshot: DataSnapshot) {
                val bikes = snapshot.children.map { dataSnapshot ->
                    dataSnapshot.getValue(Bike::class.java)!!
                }
                liveData.postValue(bikes)
            }

            override fun onCancelled(error: DatabaseError) {

            }
        })
    }

    fun updateBike(bike: Bike, bikeFirebaseKey: String) {
        val bikeValues = bike.toMap()

        REF_DATABASE_ROOT
            .child(USERS_NODE)
            .child(CURRENT_UID)
            .child(BIKES_NODE)
            .updateChildren(hashMapOf<String, Any?>(
                bikeFirebaseKey to bikeValues
            ))
    }

    fun deleteBike(bikeKey: String) {

        /** Delete bike from DB */
        REF_DATABASE_ROOT
            .child(USERS_NODE)
            .child(CURRENT_UID)
            .child(BIKES_NODE)
            .child(bikeKey).removeValue()

        /** Delete bike image from store */
        REF_STORAGE_ROOT
            .child(USERS_NODE)
            .child(CURRENT_UID)
            .child(BIKES_NODE)
            .child(bikeKey).delete()
    }

}